torch/csrc/jit/pickler.cpp

```diff
@@ -812,18 +846,53 @@ PickleOpCode Unpickler::readInstruction() {
           int64_t storage_offset = elements.at(idx++).toInt();
           std::vector<int64_t> size = tupleToIntList(elements.at(idx++));
           std::vector<int64_t> stride = tupleToIntList(elements.at(idx++));
-          double q_scale = 0.;
-          int64_t q_zero_point = 0;
+          at::Tensor result;
           if (quantized) {
-            q_scale = elements.at(idx++).toDouble();
-            q_zero_point = elements.at(idx++).toInt();
+            auto qparams_tuple = elements.at(idx++).toTuple();
+            const auto& qparams = qparams_tuple->elements();
+            auto qscheme = static_cast<at::QScheme>(qparams.at(0).toInt());
+            switch (qscheme) {
+              case at::kPerTensorAffine: {
+                double q_scale = qparams.at(1).toDouble();
+                int64_t q_zero_point = qparams.at(2).toInt();
+                result = at::_empty_affine_quantized(
+                    {0}, storage_tensor.options(), q_scale, q_zero_point);
+              } break;
+              case at::kPerChannelAffine: {
+                const auto& scales_tuple = qparams.at(1).toTuple()->elements();
+                auto scales =
+                    at::empty({scales_tuple.size()}, at::dtype<double>());
+                auto* scales_ptr = scales.data_ptr<double>();
+                for (int i = 0; i < (int)scales_tuple.size(); ++i) {
+                  scales_ptr[i] = scales_tuple[i].toDouble();
+                }
+                const auto& zero_points_tuple =
+                    qparams.at(2).toTuple()->elements();
+                auto zero_points =
+                    at::empty({zero_points_tuple.size()}, at::dtype<int64_t>());
+                auto* zero_points_ptr = zero_points.data_ptr<int64_t>();
+                for (int i = 0; i < (int)zero_points_tuple.size(); ++i) {
+                  zero_points_ptr[i] = zero_points_tuple[i].toInt();
+                }
+                std::vector<int64_t> axis;
+                for (const auto& x : qparams.at(3).toTuple()->elements()) {
+                  axis.push_back(x.toInt());
+                }
+                result = _empty_per_channel_affine_quantized(
+                    {0}, scales, zero_points, axis, storage_tensor.options());
```

Comment:
Why are all the parameters for `_empty_per_channel_affine_quantized` `Tensor`s instead of a list?

Reply:
because `scales` is float array, we don't have float array and we don't have plans to support it

Reply:
also, very theoretically, we might have multi-axis quantization in which case scales would be a multidimensional tensor. It's good to have a support for it in api, but it's not likely to appear soon. (note that in serialiation it's still fine to fold the scales to a tuple as the shape is deducible from the original tensor.